### ISSUE TITLE
Editing BCs in an already existing .feb file

### ISSUE THREAD

**Issue by jacquesbarman27** [gibbonCode/GIBBON#147]
Hello everyone! 

I am trying to use Gibbon to change the boundary condition value of an already existing .feb file. Very similar to what is done with set_mat_par_FEBIO but with BCs instead of material properties. 

Basically, I want to change the 0.015 to 0.030 but I want to use Gibbon to do this since in reality, I will be creating hundreds of .feb files with different values for the displacement. 

![image](https://user-images.githubusercontent.com/78436570/187829788-535e60ea-0180-4b5b-9a20-602bfc2f95ae.png)

I can't find a function that does this, and I am not sure how to code it. If anyone has done something similar or any idea on how to do this I would appreciate it!

**Comment by DavidBioComp** [comment]
Hi jacques,

I am encountering a similair problem, have you already been able to come up with a solution?  The import_FEB.m function can import .feb files. I however have not been able to import .feb files in a adjustable configration, as I have thus far only found files which create .feb models from scratch in the GIBBON-master folder.

**Comment by Kevin-Mattheus-Moerman** [comment]
@jacquesbarman27 apologies for the delay in getting back to you. A full and proper import of feb files is challenging. I do not have a great method for it yet. One way to process changes to the .feb file is to simply use text/string operations in MATLAB. E.g. to find a line of text and to edit it and export the new file. Below is a more complex and more efficient approach whereby we actually part the XML file using proper MATLAB-XML tools. However this, in my oppinion, always looks horrible :) this is why I made the FEBio coding in MATLAB be structure based rather than this quasi-java-matlab language. Anyway I hope it helps. 

Here is an example similar to yours, let's say we wanted to edit the `scale` value for the `bc` with the `name` attribute `prescibe_z` (my example features searching based on name, but you can edit it to use other attributes, e.g. the node_set used). 

Let's say the feb file XML for the `Boundary` stuff looks like this: 
```xml
    <Boundary>
        <bc name="fix_x" node_set="bcSupportList_X" type="fix">
            <dofs>x</dofs>
        </bc>
        <bc name="fix_y" node_set="bcSupportList_Y" type="fix">
            <dofs>y</dofs>
        </bc>
        <bc name="fix_z" node_set="bcSupportList_Z" type="fix">
            <dofs>z</dofs>
        </bc>
        <bc name="prescibe_z" node_set="bcPrescribeList" type="prescribe">
            <dof>z</dof>
            <scale lc="1">-3</scale>
            <relative>0</relative>
        </bc>
    </Boundary>
```
Then the following code finds the 4th entry based on the name `prescibe_z` (or whatever `bcPick` is set to), and then edits its `scale` to be `bcNewVal`. 

```matlab
clear; close all; clc; 

%%
fileName='/mnt/data/MATLAB/GIBBON/data/temp/tempModel.feb'

bcPick='prescibe_z';
bcNewVal=3.123456;

%% Import XML as XML object
FEB_XML = xmlread(fileName); %Read in XML object

%% Process the "Boundary" xml content

Boundary_FEB_XML = FEB_XML.getElementsByTagName('Boundary'); %Boundary level
bc_FEB_XML = Boundary_FEB_XML.item(0).getElementsByTagName('bc'); %bc level

%Find number of bc entries
numBCs=bc_FEB_XML.getLength; %Number of bc entries found

%% Find relevant bc to edit 

% Loop over the bc's and store the "name" attribute
bc_name=cell(1,numBCs); %Initialise cell to store attribute strings
for q=1:1:numBCs %loop over bc's
    bc_name{q}=bc_FEB_XML.item(q-1).getAttribute('name').toCharArray()'; %note zero indexing is used
end

%Find index of target bc
indPick=find(contains(bc_name,bcPick)); %Index of desired bc

%% Edit the bc "scale" 

%Get the "scale" level
bc_FEB_XML_scale=bc_FEB_XML.item(indPick-1).getElementsByTagName('scale');

%Change value of the boundary condition "scale"
textVal=sprintf('%6.7e,',bcNewVal); %Text to use for value, alternative: sprintf('%6.7e,',bcNewVal)
bc_FEB_XML_scale.item(0).getFirstChild.setData(textVal); %Set new text

% Export new xml file
xmlwrite(fileName,FEB_XML);
```
Does this help? I realise this is quite a complex looking bit of code. Let me know if I can help explain it further. 

P.S. Note that once you have the XML object in MATLAB you can "mess around" with it in the command window to access different components. Sometimes the _Tab_ key will help to suggest what to do with the object to. See this gif: 

![Peek 2022-09-13 15-02](https://user-images.githubusercontent.com/8392709/189922217-c3947ecb-6488-42c0-adcd-f669b365e32c.gif)

**Comment by jacquesbarman27** [comment]
Hi Kevin, thank you! 

I had created a Matlab script that read the .feb file as an array of strings. Inside the feb file, I placed dummy names for what I wanted to change (e.g. instead of having a value for the scale in a BC, I would manually write CHANGE). I would have Matlab search for the word CHANGE in the array of strings and replace it with the value I needed. Then I created a new feb file. Your code looks to work more efficiently and with less manual work.
